### ISSUE TITLE
Rsyslog with multiple remote targets using omfwd and configuration present in different files.

### ISSUE THREAD

**Issue by AbhishekRamamurthy** [rsyslog/rsyslog#5093]
### Expected behavior
Report logs to multiple omfwd targets, when omfwd targets are placed under different configuration file in /etc/rsyslog.d/1.conf 2.conf

### Actual behavior
Logs are forwarded to only one of the targets

### Steps to reproduce the behavior
Two rsyslog collectors A and B. 
Linux machine with rsylogd client running, which will report logs to A and B.

1) create two forwarding configuration under /etc/rsyslog.d/
x.conf  and y.conf
2) with x.conf reporting over TCP to server A and  y.conf reporting over UDP to server B.
3) Restart rsyslogd (to make rsyslog pick up the newly created configuration)
4) On current version only of the config file is being picked up (x.conf), if I want y.conf to work I need to remove x.conf.

x.conf
```
ubuntu@e118-NV-001:/etc/rsyslog.d$ cat 60-remote-syslog-forwarder.conf

$DefaultNetstreamDriver gtls
$ActionSendStreamDriverMode 1
$ActionSendStreamDriverAuthMode x509/certvalid
$DefaultNetstreamDriverCAFile /etc/tls/rsyslog/rsyslog-ca.pem
$DefaultNetstreamDriverCertFile /etc/tls/rsyslog/rsyslog-crt.pem
$DefaultNetstreamDriverKeyFile /etc/tls/rsyslog/rsyslog-key.pem


ruleset(name="RuleSet0") {
action(type="omfwd"
   queue.type="LinkedList"
   queue.filename="Forwarder0"
   queue.maxDiskSpace="10485760"
   queue.discardSeverity="3"
   queue.timeoutEnqueue="0"
   action.resumeRetryCount="3"
   Protocol="tcp"
   Target="2.3.4.5"
   Port="8012"

  StreamDriver="gtls"
  StreamDriverMode="1"
  StreamDriverAuthMode="x509/certvalid"


  )
}
input(type="imfile"
  File="/var/log/syslog"
  Tag="syslog"
  Ruleset="RuleSet0")
input(type="imfile"
  File="/var/log/auth.log"
  Tag="auth.log"
  Ruleset="RuleSet0")
```

y.conf
```
ruleset(name="RuleSet9") {
action(type="omfwd"
   queue.type="LinkedList"
   queue.filename="Forwarder9"
   queue.maxDiskSpace="10485760"
   queue.discardSeverity="3"
   queue.timeoutEnqueue="0"
   action.resumeRetryCount="3"
   Protocol="udp"
   Target="5.6.7.8"
   Port="9012"


  )
}
input(type="imfile"
  File="/var/log/syslog"
  Tag="syslog"
  Ruleset="RuleSet9")
input(type="imfile"
  File="/var/log/auth.log"
  Tag="auth.log"
  Ruleset="RuleSet9")
```

### Environment
ubuntu 18.04

Issue observed on.

```
ubuntu@e118-NV-001:/etc/rsyslog.d$ rsyslogd -v
rsyslogd  8.2212.0 (aka 2022.12) compiled with:
	PLATFORM:				x86_64-pc-linux-gnu
	PLATFORM (lsb_release -d):
	FEATURE_REGEXP:				Yes
	GSSAPI Kerberos 5 support:		No
	FEATURE_DEBUG (debug build, slow code):	No
	32bit Atomic operations supported:	Yes
	64bit Atomic operations supported:	Yes
	memory allocator:			system default
	Runtime Instrumentation (slow code):	No
	uuid support:				Yes
	systemd support:			Yes
	Config file:				/etc/rsyslog.conf
	PID file:				/var/run/rsyslogd.pid
	Number of Bits in RainerScript integers: 64

See https://www.rsyslog.com for more information.
```

last known working release.
```
ubuntu@ip-172-31-3-122:/etc$ rsyslogd -v
rsyslogd 8.32.0, compiled with:
	PLATFORM:				x86_64-pc-linux-gnu
	PLATFORM (lsb_release -d):
	FEATURE_REGEXP:				Yes
	GSSAPI Kerberos 5 support:		Yes
	FEATURE_DEBUG (debug build, slow code):	No
	32bit Atomic operations supported:	Yes
	64bit Atomic operations supported:	Yes
	memory allocator:			system default
	Runtime Instrumentation (slow code):	No
	uuid support:				Yes
	systemd support:			Yes
	Number of Bits in RainerScript integers: 64

See http://www.rsyslog.com for more information.
```

**Comment by davidelang** [comment]
you are misunderstanding the configuration language, what file something is in 
doesn't matter, so the fact that one file has inputs and so does another file 
doesn't matter

when rsyslog starts up, it goes through and finds all the startup-time commands, 
this is all of the input and module statements (and a few others), it sets those 
up. In your case, you define two inputs for the same file, but with different 
rulesets. Rsyslog does not allow you to define two inputs for the same file.

so only one of the input commands gets used, I think that the last that it sees 
wins, but that may be undefined behavior that happens to work that way.

As a result, only one of your rulesets gets exectuted, and since each ruleset 
only has one omfwd, it's only sent one place.

If you create a ruleset that has multiple omfwd actions in it, it will send it 
multiple places.

you don't have to use the default include that most distros have 
(/etc/rsyslog.d/*.conf), you can make whatever include you want, so you could 
include /etc/rsyslog.d/syslog*.conf and /etc/rsyslog.d/auth*.conf.d

It's also a really bad idea to set imfile to read the files that rsyslog will be 
writing out by default (it tends to lead to infinate loops). If you are creating 
those files in a different instance of rsyslog, why not do the forwarding there 
instead?

feel free to ask for more explinations, or to give more explination about the 
problem you are trying to solve with these duplicate input statements.

David Lang

**Comment by AbhishekRamamurthy** [comment]
@davidelang Thanks for the response.

1) we are using only one rsyslogd instance as client, same instance is used for populating syslog and auth.log as well as reporting the files to remote server.
2) We have multiple log rsyslogd server instances on different machine, each machine may have its own TLS certificate or can run on UDP.
3) With this we were able to separate out configuration for each remote server and its certificate in a different file.
4) We were using the above approach which worked in 8.32, to handle this use case. Is there any recommended approach to handle multiple TLS end points with each having its own certs ?

**Comment by davidelang** [comment]
On Sat, 25 Feb 2023, Abhishek Ramamurthy wrote:

> 1) we are using only one rsyslogd instance as client, same instance is used for populating syslog and auth.log as well as reporting the files to remote server.

In that case, you really don't want to write to the file and read it again, just 
add your send to the remote system in the same filter that writes it to a file.

you can do something like

auth.* {
   /var/log/auth.log
   @@1.2.3.4 #or any other action
   $include /etc/rsyslog.d/omfwd/*.conf
}

> 2) We have multiple log rsyslogd server instances on different machine, each machine may have its own TLS certificate or can run on UDP.
> 3) With this we were able to separate out configuration for each remote server and its certificate in a different file.

as I show above, you can have an include, and then have the per system files in 
an /etc/rsyslog.d/omfwd as needed

> 4) We were using the above approach which worked in 8.32, to handle this use case. Is there any recommended approach to handle multiple TLS end points with each having its own certs ?

multiple endpoints with their own certs each need their own action() clause (and 
8.32 didn't support more than one set of certs, that's a very recent addition). 
That's now where you are going in an odd direction, reading the file that you 
are simultaniously writing and reading the same file multiple times are what you 
are doing that's odd.

We'll have to see Rainer's comment, we try really hard not to break old configs, 
so this may be something we can make work again, but it's really a questionable 
approach.

David Lang

**Comment by AbhishekRamamurthy** [comment]
@davidelang thanks for the response. I will try out your recommendation and update here. But is there a chance of keeping the 8.32 behavior consistent with 8.22xx ?

**Comment by rgerhards** [comment]
I do not know if 8.32 really behaves other than newer versions, but if so: this is probably a side-effect of some non-documented behavior. It was always recommended against writing files and reading them again. So, no, this is nothing we will even try to address.

**Comment by rgerhards** [comment]
Withe "non-documented" I mean with a behavior that was never guaranteed to work in this way. Reading a file we write ourselfs is not well defined. Also when other processes write files in certain ways, surprises are always possible. imfile is the best solution for a bad situation, the core idea is to grab messages right at the source API.

**Comment by davidelang** [comment]
I think the actual issue here is having multiple input() clauses reading the 
same file. The fact that it's also a file that he is writing out is just an 
additional complication.

thinking about it, this could be because inotifiy will only notify us once for 
the file, if he switched to polling mode (less efficient), that may work, and 
the switch from polling to inotify by default could easily be a change in the 5+ 
years since 8.32 was released

Is there anything else in the code that you can think of that would break with 
multiple inputs on the same file?

still undocumented behavior though

David Lang

**Comment by AbhishekRamamurthy** [comment]
@davidelang @rgerhards 

If remote targets are included at filter, will this serialize the log ?, as In a new line will be added only after current line is reported all remote targets ?

auth.* {
   /var/log/auth.log
   $include /etc/rsyslog.d/omfwd/*.conf # has multiple conf files with different remote target defined as action
}

**Comment by davidelang** [comment]
Just the opposite, it will write to /var/log/auth.log, then send to all the 
remote targets.

if you start rsyslog with -o /path/to/file it will create a config in 'file' 
that is what it sees after all the includes are processed

David Lang

**Comment by linzhuorong** [comment]
This issue mentions some reference information:
https://github.com/rsyslog/rsyslog/issues/4897

**Comment by amg7707** [comment]
> @davidelang @rgerhards
> auth.* { /var/log/auth.log $include /etc/rsyslog.d/omfwd/*.conf # has multiple conf files with different remote target defined as 
action }

Hello,

How would it be in RainerScript ?
I want to write a haproxylog and then to send it to two targets (one udp/514, one tcp/6514).

Thanks for you help.

Regards.

**Comment by davidelang** [comment]
amg7707 wrote:

>> @davidelang @rgerhards
>> auth.* { /var/log/auth.log $include /etc/rsyslog.d/omfwd/*.conf # has multiple conf files with different remote target defined as
> action }
>
> Hello,
>
> How would it be in RainerScript ?
> I want to write a haproxylog and then to send it to two targets (one udp/514, one tcp/6514).

I'm not understanding the question, it could be as simple as just putting two 
action() statements in, one to each destination.

note that rsyslog doesn't care if things are in one file or if every line is in 
a separate file. Rsyslog reads all the files in and combines them before it 
starts interpreting anything.

David Lang

**Comment by davidelang** [comment]
yep, it really is that simple

David Lang

**Comment by rgerhards** [comment]
Apologies for the late follow-up. This is probably no longer relevant for the original reporter, but I am adding the answer for people who find this issue through search. We do not lightly close old issues just because of age; we are working through the backlog and closing them when the old thread has a clear answer.

The important point is that files in `/etc/rsyslog.d/` are merged into one effective configuration before interpretation. File placement does not create independent pipelines. Defining multiple `imfile` inputs for the same file and trying to route them independently is not a supported way to send the same content to multiple destinations, and reading a file that rsyslog itself writes can also create undefined feedback-loop behavior.

The recommended pattern is to have one input/ruleset for the source and put the multiple output actions under that same rule or ruleset. If you want to keep target-specific snippets in separate files, include snippets that define actions under the same rule path rather than duplicate the input. The reporter later confirmed that this recommendation worked, so I am closing the issue as answered/configuration guidance.
